src/Test/CloseFileHandles.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE CPP #-}
+module Test.CloseFileHandles(main) where
+
+import Development.Shake
+import Development.Shake.FilePath
+import Test.Type
+import Control.Monad.Extra
+import System.Exit
+import System.IO
+
+#ifdef mingw32_HOST_OS
+
+main = return () -- don't know how to do this on windows
+
+#else
+
+import System.Posix.IO
+
+main = testBuild test $ do
+    let helper = toNative $ "helper/close_file_handles_helper" <.> exe
+    let name !> test = do want [name]
+                          name ~> do need ["helper/close_file_handles_helper" <.> exe]; test
+
+    let helper_source = unlines
+            ["import System.Environment"
+            ,"import System.Posix.IO"
+            ,"import System.IO"
+            ,"import System.Exit"
+            ,""
+            ,"main = do"
+            ,"  args <- getArgs"
+            ,"  case args of"
+            ,"    [fdString] -> do"
+            ,"       handle <- fdToHandle (read fdString)"
+            ,"       hClose handle"
+            ,"       exitSuccess"
+            ,"    _ -> do"
+            ,"      progName <- getProgName"
+            ,"      hPutStrLn stderr (\"usage: \" ++ progName ++ \" <file descriptor number>\\n    tries closing the file descriptor number\\n    exits successful, if the file descriptor was open\")"
+            ,"      exitWith (ExitFailure 3)"]
+    "close_file_handles_helper.hs" %> \out -> do
+        need ["../../src/Test/CloseFileHandles.hs"]
+        writeFileChanged out helper_source
+    ["helper/close_file_handles_helper"<.>exe, "close_file_handles_helper.hi", "close_file_handles_helper.o"] &%> \_ -> do
+        need ["close_file_handles_helper.hs"]
+        cmd "ghc --make" "close_file_handles_helper.hs -o helper/close_file_handles_helper" 
+
+    "defaultbehaviour" !> do
+        withTempFile $ \file -> do
+            h <- liftIO $ openFile file AppendMode  -- should probably be a bracket in Action () how to do that?
```

Comment:
It was impossible. I just added `actionBracket` that should be perfect for this. Can you try converting the code over and report back?